lib/stockboy/providers/ftp/sftp.rb

```diff
@@ -0,0 +1,66 @@
+require 'net/sftp'
+
+class SFTPAdapter
+  attr_accessor :client
+
+  def initialize(provider)
+    @provider = provider
+  end
+
+  def open
+    result = nil
+    Net::SFTP.start(@provider.host, @provider.username, password: @provider.password) do |sftp|
+      puts 'open connection'
+      @client = sftp
+      result = yield self
+      puts 'close connection'
+    end
+    result
+  end
+
+  def chdir(directory)
+    @file_dir = directory.nil? || directory.empty? ?  '/' : normalize_file_dir(directory)
+  end
+
+  def list_files
+    client.dir.entries(@file_dir).map { |e| e.name }
+  end
+
+  def delete(file_name)
+    client.remove!("#{@file_dir}#{file_name}")
+  end
+
+  def download(file_name)
+    file = client.file.open("#{@file_dir}#{file_name}")
+    data = file.gets
+    file.close
```

Comment:
This is the only place that `file.close` gets called, but there are open file handles in all the other methods too. Maybe worth a helper method for this. Note that `file.open` also takes a block so it can handle closing for you:

https://github.com/net-ssh/net-sftp/blob/master/lib/net/sftp/operations/file_factory.rb#L39